### PR TITLE
feat: get branch name without calling git

### PR DESCRIPTION
**Pull request by grtcdr**
Getting the branch name previously required calling `git branch --show-current`. 

This PR addresses that by parsing `.git/HEAD` directly. Further testing is required to observe how this new change behaves when the user is in a detached HEAD. I'll report back!

EDIT: Detached HEAD successfully shows the entire commit hash, I'll limit it to the first 7 characters.

EDIT 2: Done :smile:

### CODE REVIEW

**Review by GitOpsBot — CHANGES_REQUESTED** [review]
# GoGitOps Review

## Grade: A+ (92.2%)

### Files: 6

### Issues: 2

### gofmt: 83%

```diff
- This PR will be rejected until gofmt is 100%
```
	cmd/getters.go
		Line 1: warning: file is not gofmted with -s (gofmt)
### go_vet: 100%

### gocyclo: 100%

### golint: 83%

	cmd/root.go
		Line 29: warning: exported function Execute should have comment or be unexported (golint)
### license: 100%

### ineffassign: 100%

### misspell: 100%




This report was generated using [GoGitOps](https://gogitops.cf).

**Review by GitOpsBot — COMMENTED** [review]
# GoGitOps Review

## Grade: A+ (98.0%)

### Files: 6

### Issues: 1

### gofmt: 100%

### go_vet: 100%

### gocyclo: 100%

### golint: 83%

	cmd/root.go
		Line 29: warning: exported function Execute should have comment or be unexported (golint)
### ineffassign: 100%

### license: 100%

### misspell: 100%




This report was generated using [GoGitOps](https://gogitops.cf).

**Review by GitOpsBot — COMMENTED** [review]
# GoGitOps Review

## Grade: A+ (98.0%)

### Files: 6

### Issues: 1

### gofmt: 100%

### go_vet: 100%

### gocyclo: 100%

### golint: 83%

	cmd/root.go
		Line 29: warning: exported function Execute should have comment or be unexported (golint)
### license: 100%

### ineffassign: 100%

### misspell: 100%




This report was generated using [GoGitOps](https://gogitops.cf).

**Comment by gennaro-tedesco** [comment]
Good evening and thank you very much for contributing to `archimede`! I am happy to test the changes! Could you please explain why it is better to obtain the current branch by `.git/HEAD` rather than `git branch --show-current`: is it because it address corner cases (like detached HEAD, for instance, that I had not considered) or perhaps to avoid calls to `exec.Command`?

**Comment by grtcdr** [comment]
> Good evening and thank you very much for contributing to `archimede`

My pleasure :heart: 

> or perhaps to avoid calls to exec.Command?

This. I don't know how fast Go is at calling external commands, but I presume it can't be faster than reading a file.

**Comment by grtcdr** [comment]
Upon further research, it appears that there is a [Go implementation of Git](https://github.com/go-git/go-git) which we can use to the do all the heavy lifting for us, eliminating the need for `exec.Command`. What do you think?

**Comment by gennaro-tedesco** [comment]
> Upon further research, it appears that there is a [Go implementation of Git](https://github.com/go-git/go-git) which we can use to the do all the heavy lifting for us, eliminating the need for `exec.Command`. What do you think?

I am having a look and it seems that it should address quite well the cases that are present in the current code: in particular [this](https://github.com/go-git/go-git/blob/c785af3f4559ebac52c42f12d17cb118aac383ad/repository.go#L1391) should fetch the reference that HEAD is pointing to.  Another use case could be to directly report whether we are in a git repo [here](https://github.com/gennaro-tedesco/archimede/blob/3fc3707e383af7135769ccc2f47d565a104fa61b/cmd/getters.go#L139) without having to again invoke `exec.Command` (though I have not found the corresponding bit in the `go-git` yet).

**Comment by gennaro-tedesco** [comment]
> This. I don't know how fast Go is at calling external commands, but I presume it can't be faster than reading a file.

In reference to the above I have tested the difference between both branches with [hyperfine](https://github.com/sharkdp/hyperfine), with the following results:
```
# your new branch
Benchmark 1: archimede
  Time (mean ± σ):      98.2 ms ±   1.2 ms    [User: 43.5 ms, System: 50.5 ms]
  Range (min … max):    96.2 ms … 101.8 ms    29 runs

# the current master branch
Benchmark 1: archimede
  Time (mean ± σ):     130.5 ms ±   1.4 ms    [User: 57.1 ms, System: 64.2 ms]
  Range (min … max):   128.4 ms … 134.0 ms    22 runs
```
so avoiding calls to `exec.Command` does indeed improve execution time. Probably using a native solution for `isGitStatus` should bring this down even further.

**Comment by grtcdr** [comment]
I forgot about the refs!

So should we walk the directories of `ref:`, component by component, and only report back the last element (the branch name) if it really exists?

> so avoiding calls to exec.Command does indeed improve execution time.

Loving the benchmarks already, we could probably even get it down to 30 ms, with our own implementation, go-git's (high-level) or [git2go](https://github.com/libgit2/git2go)'s (low-level).

**Comment by gennaro-tedesco** [comment]
> I forgot about the refs!

In what context would it be useful to check `.git/refs`, for the cases at hand? 

> with our own implementation, go-git's (high-level) or [git2go](https://github.com/libgit2/git2go)'s (low-level).

I would approach this step by step: namely the current solution (your branch) already offers good performance enhancement (with the trick of reading the `.git/HEAD`), we could go ahead and merge this already and then with ease look into which one between `go-git` and `git2go` makes the whole experience smoother.

Any equivalent trick to circumvent 
```
func isGitRepo() bool {
	_, err := exec.Command("sh", "-c", "git rev-parse --is-inside-work-tree 2>/dev/null").Output()
```
? Would in this case checking for the mere existence of the `.git` folder suffice (though this is more or less what `git rev-parse` does, namely looking for commit id of the current head and if not existing throwing an error)?

File: cmd/getters.go [hunk 1]
```diff
@@ -145,8 +145,21 @@ func isGitRepo() bool {
 }
 
 func getGitStatus() map[string]string {
-	gitBranch, eb := exec.Command("sh", "-c", "git branch --show-current").Output()
-	if eb != nil {
+	branch, eb := os.ReadFile(".git/HEAD")
```
Comment (gennaro-tedesco) [inline cmd/getters.go:148]: while testing I have noticed that this fails in case you are within a sub-folder of the project (since the path `.git/` does not point to the root dir), whereas these cases are handled automatically by the `git branch` parser. We need to understand how to solve this case (probably worth looking into `go-git` for it, at this point).

**Review by gennaro-tedesco — CHANGES_REQUESTED** [review]

File: cmd/getters.go (hunk 1 above)
Comment (grtcdr) [reply cmd/getters.go:148]: We can walk back up until we hit `.git`, right?

Comment (grtcdr) [reply cmd/getters.go:148]: Would you like me to leverage `go-git` instead of reinventing the wheel?

Comment (grtcdr) [reply cmd/getters.go:148]: I'm really not sure `go-git` is a good choice though, looking through its issue tracker leads me to believe it's not being taken care of. `git2go` seems like a better, slightly lower-level option.

**Comment by grtcdr** [comment]
> we could go ahead and merge this already and then with ease look into which one between go-git and git2go makes the whole experience smoother.

~~Sure, as you like!~~ This PR is still making decisions on presumptions, which we should take care of before merging.

> Would in this case checking for the mere existence of the .git folder suffice

This won't suffice. `git rev-parse --is-inside-work-tree 2>/dev/null` may also fail in cases where the repository was initiated with the `--bare` option.

**Comment by grtcdr** [comment]
Unfortunately, I can't `go get github.com/libgit2/git2go/v33` because it results in:

```
error "Invalid libgit2 version; this git2go supports libgit2 between v1.3.0 and v1.3.0"
``` 

My libgit2 version is `1:1.4.2`, so that's why. (I'm going to have to wait until they bump `libgit2`)

I see no mentions of the libgit2 version being used by `go-git`. *Ugh!*.

**Comment by gennaro-tedesco** [comment]
I will have a look at `git2go` myself too, these days, to see how it can help. To summarise what we discussed so far the problems we need to address to approve the PR are:

1. check if we are in a git repository at all in the first place
2. correctly parse branch name from any location within the working directory (even sub-folders)
3. correctly address the case of bare repository 

ideally avoiding calls to `exec.Command`, rather either using high-level libraries (`go-git`, `git2go`) or by looking up `.git` folders and files.

### status quo
- the `master` branch uses `exec.Command`, works for 1. and 2. but fails for 3.
- the proposed PR addresses 2. but fails in sub-folders.

**Comment by grtcdr** [comment]
Thanks for the recap.

One thing to note about bare repositories, is that they don't contain a workspace, only the repository. So we can safely fail in that case.

**Comment by gennaro-tedesco** [comment]
I had a look at possible alternatives (in particular `git2go`) but could not find any intuitive API that addresses 1. and 2. in a clean way. I will continue investigate to see perhaps how (and if) this is done in other more famous repositories that parse the git status for high level projects.
